File: frontend/src/services/admin/sale.service.ts
import api from '@/services/api';

export interface SaleItem {
    id: number;
    name: string;
    discountPercent: number;
    couponCode?: string;
    startTime?: string;
    endTime?: string;
    status: 'draft' | 'active' | 'ended';
    active: boolean;
    productIds: number[];
    createdAt: string;
}

export const adminSaleService = {
    getAll: async (): Promise<SaleItem[]> => {
        const res = await api.get('/admin/sales');
        return res.data?.data || [];
    },
    getById: async (id: number): Promise<SaleItem> => {
        const res = await api.get(`/admin/sales/${id}`);
        return res.data?.data;
    },
    create: async (data: Partial<SaleItem>): Promise<SaleItem> => {
        const res = await api.post('/admin/sales', data);
        return res.data?.data;
    },
    update: async (id: number, data: Partial<SaleItem>): Promise<SaleItem> => {
        const res = await api.put(`/admin/sales/${id}`, data);
        return res.data?.data;
    },
    delete: async (id: number): Promise<void> => {
        await api.delete(`/admin/sales/${id}`);
    },
};

export const getActiveSales = async (): Promise<SaleItem[]> => {
    try {
        const res = await api.get('/sales/active');
        return res.data?.data || [];
    } catch {
        return [];
    }
};

File: backend/lili-fashion-api/src/main/java/com/lilifashion/module/sale/service/SaleService.java
package com.lilifashion.module.sale.service;

import com.lilifashion.common.exception.AppException;
import com.lilifashion.module.product.entity.Product;
import com.lilifashion.module.product.repository.ProductRepository;
import com.lilifashion.module.sale.dto.SaleDto;
import com.lilifashion.module.sale.dto.SaleRequest;
import com.lilifashion.module.sale.entity.Sale;
import com.lilifashion.module.sale.repository.SaleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SaleService {

    private final SaleRepository saleRepository;
    private final ProductRepository productRepository;

    @Transactional(readOnly = true)
    public List<SaleDto> getAll() {
        return saleRepository.findAll().stream().map(SaleDto::from).collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public SaleDto getById(Long id) {
        return SaleDto.from(saleRepository.findById(id)
                .orElseThrow(() -> AppException.notFound("Sale")));
    }

    @Transactional(readOnly = true)
    public List<SaleDto> getActiveSales() {
        return saleRepository.findActiveSales(LocalDateTime.now()).stream()
                .map(SaleDto::from).collect(Collectors.toList());
    }

    @Transactional
    public SaleDto create(SaleRequest req) {
        Sale sale = Sale.builder()
                .name(req.getName())
                .discountPercent(req.getDiscountPercent())
                .couponCode(req.getCouponCode())
                .startTime(req.getStartTime())
                .endTime(req.getEndTime())
                .status(req.getStatus() != null ? Sale.SaleStatus.valueOf(req.getStatus().toUpperCase()) : Sale.SaleStatus.DRAFT)
                .build();

        if (req.getProductIds() != null && !req.getProductIds().isEmpty()) {
            List<Long> ids = req.getProductIds();
            List<Product> products = productRepository.findAllById((Iterable<Long>) ids);
            sale.setProducts(products);
        }

        return SaleDto.from(saleRepository.save(sale));
    }

    @Transactional
    public SaleDto update(Long id, SaleRequest req) {
        Sale sale = saleRepository.findById(id).orElseThrow(() -> AppException.notFound("Sale"));

        if (req.getName() != null) sale.setName(req.getName());
        if (req.getDiscountPercent() != null) sale.setDiscountPercent(req.getDiscountPercent());
        if (req.getCouponCode() != null) sale.setCouponCode(req.getCouponCode());
        if (req.getStartTime() != null) sale.setStartTime(req.getStartTime());
        if (req.getEndTime() != null) sale.setEndTime(req.getEndTime());
        if (req.getStatus() != null) sale.setStatus(Sale.SaleStatus.valueOf(req.getStatus().toUpperCase()));
        if (req.getProductIds() != null) {
            List<Long> ids = req.getProductIds();
            sale.setProducts(productRepository.findAllById((Iterable<Long>) ids));
        }

        return SaleDto.from(saleRepository.save(sale));
    }

    @Transactional
    public void delete(Long id) {
        saleRepository.deleteById(id);
    }
}
